### DISCUSSION TITLE
Transform Strings to Non-strings

### DISCUSSION THREAD

**Discussion by Narinderg** [babel/babel#13334] [Q&A] [answered]
Dear All,

We have a requirement to transform strings like "<#lang keyname>", '<#lang keyname>' to an expression like meta.find("keyname").
If we using StringLiteral visitor pattern, it doesn't take the starting and ending quotes/doublequotes.
What can be the possible solution here? Please help.

**Comment by fedeci** [comment, marked as answer]
If I understood correctly your question you are trying to preserve quotes or double quotes, right?
In that case you may want to use the
`extra.raw` property of the `StringLiteral` node.

**Reply by Narinderg** [reply]
Thanks @fedeci for your response. I actually tried that as well. 
path.node.raw also returned without quotes (just like path.node.value). path.node.extra actually returned undefined.
Below is the code of my plugin. Will really appreciate any further assistance here. Thanks.

const META_HELPERS = "metaHelpers";

const transformations = [
  {
    search: /(`|\"|')<#lang (.*?)(,.*?)?#>(`|\"|')/,
    replace: (match, p1, p2, p3, p4) => {
      return `${META_HELPERS}.getLabel("${p1.trim()}"${
        p2 ? p2.trim() : ""
      })`;
    },
  },
  {
    search: "(`|\"|')<#settings (.*?)(,.*?)?#>(`|\"|')",
    replace: (match, p1, p2, p3, p4) => {
      return `${META_HELPERS}.getSettings("${p1.trim()}"${
        p2 ? p2.trim() : ""
      })`;
    },
  },
];


function visitor(path, state) {
 
  transformations.map((opt) => {
  console.log("value", path.node.extra); //gives undefined
  const isStringLiteral = path.node.type === "StringLiteral";

    if (isStringLiteral) {
      if (path.node.raw && path.node.raw.match(opt.search)) { // tried both path.node.raw as well as path.node.value
        console.log("value", path.node.value);
        path.node.value = path.node.raw.replace(opt.search, opt.replace);
        console.log("value after", path.node.value);
        return;
      }
      return;
    }

  });
}

module.exports = function logger({
  types: t
}) {
  return {
    name: 'logger',
    visitor: {
      StringLiteral: visitor,
    },
  };
}

**Reply by fedeci** [reply]
`path.node.raw` is `undefined`, because the raw property is not defined directly in the node, but in his extra object. You have to use `path.node.extra.raw` to access it.
Do you simply want your string to be wrapped inside `meta.find(STRING)`?

**Reply by Narinderg** [reply]
My requirement is to actually transform the instances of "<#lang keyname>" to something like meta.find("keyname"). So the whole string (including quotes) needs to be transformed to that statement. 

For me when I am running the above plugin path.node.raw is actually returning the String same as path.node.value. path.node.extra is returning undefined.

**Reply by fedeci** [reply]
This should work, but it requires a check to test if the path should be replaced.
```js
module.exports = function logger({ types: t }) {
return {
	name: 'logger',
	visitor: {
            StringLiteral(path) {
                  // check if you should replace
                  const replacement = t.callExpression(t.memberExpression(t.Identifier("meta"), t.Identifier("find")), [t.stringLiteral(path.node.value)])
                  path.replaceWith(replacement)
                  path.skip()
            }
	},
};
}
```

**Reply by Narinderg** [reply]
Thanks @fedeci. I was able to make it work using the pointers given in your answer.
